the-collab-lab/tcl-57-smart-shopping-list
# Sort and display items in order of how soon they need to be purchased

## Description



This PR implements code that renders grocery list items in order of how soon the user is predicted to need to purchase each item based on the their purchasing behavior. Sorting grocery lists by purchasing urgency allows users to more easily identify what they may need to buy at any given point in time and helps them better plan their shopping trips, resulting in a better user experience. 

The sorting is handled by the `comparePurchaseUrgency` function we implemented. It:
- Accepts an unsorted list
- Filters each unsorted item into active and inactive categories and appends an urgency property to each item object. Active items fall into one of four urgency groups (`overdue`, `soon`, `kind of soon`, and `not soon`) while all inactive items fall into a single group (`inactive`). 
    - **active items**: items purchased less than 60 days ago
        - `overdue`: items whose projected purchase date has elapsed, but less than 60 days ago
        - `soon`: items projected to be purchased in 7 days or less
        - `kind of soon`: items projected to be purchased between 7 and 21 days
        - `not soon`: items projected to be purchased in 21 days or more 
    - **inactive items**: items purchased more than 60 days ago _and_ whose projected next-purchase-date has elapsed
        - `inactive`: all inactive items
- Sorts active and inactive items in ascending order of days until the projected next-purchase-date and breaks any ties alphabetically. 
- Returns a joined list with sorted active items at the front and sorted inactive items at the end. 

This function gets called inside the `List` component. The returned sorted list is then mapped over and a `ListItem` is created for each of its items. To communicate each item's urgency to the end user, we leveraged the urgency property that was appended to each item object in the `comparePurchaseUrgency` function to render plain text describing the urgency alongside a representative colored dot. 

## Related Issue
Closes #12

## Acceptance Criteria


### A stretch goal

## Type of Changes
|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|  ✓  | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

### Before



The user's grocery list was rendered unsorted. 

<img width="2160" alt="a" src="https://github.com/the-collab-lab/tcl-57-smart-shopping-list/assets/9102330/82f23220-2dae-4bd4-b271-6187c38217c4">


### After

The user's grocery list is rendered sorted ordered by active/inactive status, days until next purchase date, and alphabetically when necessary. 

<img width="2160" alt="b" src="https://github.com/the-collab-lab/tcl-57-smart-shopping-list/assets/9102330/32082456-64f6-420b-89b2-5c53bf58d324">




## Testing Steps / QA Criteria

- Visit the preview link for this PR: https://tcl-57-smart-shopping-list--pr28-jg-es-sort-items-by-1t0qhc9i.web.app/
- To test that items have a purchase urgency indicator in the UI and that items are listed in order of purchase urgency (i.e. sorted by number of days until `dateNextPurchased`):
    - Start a new list and add three items with different purchase urgency (i.e. one item should be purchased `soon`, one `kind of soon` and one `not soon`).
    - On the `List` view, ensure that each item’s urgency is indicated by a colored dot and plain text stating the purchase urgency.
    - Check that the items appear in order of `soon`, `kind of soon` and `not soon`.
- To test that items with the same number of days until `dateNextPurchased` are sorted alphabetically by item name:
    - Add a few more items which start with a different letter and have the same purchase urgency. Ex: “apples” - `kind of soon`, “bananas” - `kind of soon`, “cherries” - `kind of soon`.
    - On the `List` view, check that the items with the same purchase urgency are listed alphabetically.
- To test that overdue items are flagged `overdue` and appear at the top of the list:
    - Go to Firestore and change one of the item’s `dateLastPurchased` to a date between 60 days ago and today. Change the `dateNextPurchased` to a date before today.
    - On the `List` view, check that the item is flagged `overdue` and appears at the top of the list.
- To test that inactive items are sorted separately and appear last in the list:
    - Go to Firestore and change one of the item’s `dateLastPurchased` to more than 60 days ago. Change the `dateNextPurchased` to a date before today.
    - On the `List` view, check that the item is flagged `inactive` and appears at the bottom of the list.
